### PR TITLE
Sheffield| 26-ITP-Jan| Mona-Eltantawy | Sprint 2 | Sprint 2 course-work tasks

### PR DESCRIPTION
**Pull request by Mona-Eltantawy**
Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

### CODE REVIEW

File: Sprint-2/debug/author.js [hunk 1]
```diff
@@ -11,6 +11,17 @@ const author = {
...
+
+let text = " ";
+for (let x in author) {
+  text+= author[x] + " ";
+};
+console.log(text);
+
+// or: 
+const values = Object.values(author);
+console.log(values);
```
Comment (cjyuan) [inline Sprint-2/debug/author.js:26]: The original code attempts to output the value one by one on a new line.
Can you rewrite your code to also output all the property values one by one on a new line?

File: Sprint-2/implement/contains.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
-function contains() {}
+function contains(obj, key) {
+    return key in obj;
+}
```
Comment (cjyuan) [inline Sprint-2/implement/contains.js:3]: Consider the following two approaches for determining if an object contains a property:
```
  let obj = {}, propertyName = "toString";
  console.log( propertyName in obj );                // true
  console.log( Object.hasOwn(obj, propertyName) );   // false
```
Which of these approaches suits your needs better?
For more info, you can look up `JS "in" operator vs Object.hasOwn`.

File: Sprint-2/implement/contains.test.js [hunk 3]
```diff
@@ -20,16 +20,35 @@ as the object doesn't contains a key of 'c'
...
+    expect(contains(obj ,"d")).toBe(false);
+});
 
 // Given invalid parameters like an array
 // When passed to contains
 // Then it should return false or throw an error
+test ("contains with invalid parameters return false", () => {
+    const obj = ['a','b','c','d'];
+    expect(contains(obj , "a")).toBe (false);
+});
```
Comment (cjyuan) [inline Sprint-2/implement/contains.test.js:51]: This test cannot yet confirm that the function correctly returns false when the first argument is an array. 
This is because `contains(['a','b','c','d'], "a")` could also return `false` simply because "a" is not a key of the array.

**Arrays are objects, with their indices acting as keys**. A proper test should use a valid
key to ensure the function returns `false` specifically because the input is an array, not because the key is missing.

Besides array, you should also test other invalid parameters such as `null`, `undefined`, strings, to ensure your function works correctly.

File: Sprint-2/implement/tally.js [hunk 4]
```diff
@@ -1,3 +1,12 @@
-function tally() {}
+function tally(items) {
+    if (!Array.isArray(items)) {
+        throw new Error("Input must be an array");
+    }
+    const result ={ };
+    for (let item of items){
+        result[item]=(result[item] || 0)+1;
+    }
```
Comment (cjyuan) [inline Sprint-2/implement/tally.js]: Does the following function call returns the value you expect?
```
tally(["toString", "toString"]);
```

Suggestion:
- Look up an approach to create an empty object with no inherited properties, or
- use `Object.hasOwn()`

**Review by cjyuan — COMMENTED** [review]
Some of the code is not consistently formatted.

Have you installed the **prettier** VSCode extension and enabled "Format on save/paste" on VSCode,
as recommended in 
https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/blob/main/readme.md
?

---
If you have enabled "Format on save" but it is not working, it is likely that you haven't assign a formatter for JS file. This could happen if you have zero or multiple extensions that can format .js file.

If you have installed "Prettier" extension. To assign it as the formatter of JS code, you can try:
1. Use "Format document" to format the JS file. Sometimes, VSCode will ask you to choose a formatter, and you can manually select "Prettier".
2. Edit `settings.json` and set Prettier as the default formatter for JS.
See: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

File: Sprint-2/debug/author.js (hunk 1 above)
Comment (Mona-Eltantawy) [reply Sprint-2/debug/author.js:26]: I fixed it.

File: Sprint-2/implement/contains.js (hunk 2 above)
Comment (Mona-Eltantawy) [reply Sprint-2/implement/contains.js:3]: I understand your point her but the function used still give the required output, so I'm not sure if I have to change it to `Object.hasOwn( )` method or to keep it.

Comment (cjyuan) [reply Sprint-2/implement/contains.js:3]: Have you found out the difference between `key in object` and `Object.hasOwn(object, key)`?

As long as you know their differences, you can use either approach in your implementation because the spec is not clear exactly how the function should behave.

File: Sprint-2/implement/contains.test.js (hunk 3 above)
Comment (Mona-Eltantawy) [reply Sprint-2/implement/contains.test.js:51]: I added some more invalid parameters to the test cases.

File: Sprint-2/debug/author.js [hunk 5]
```diff
@@ -11,6 +11,17 @@ const author = {
...
-  console.log(value);
-}
+// for (const value of author) {
+//   console.log(value);
+// }
+
+let text = " ";
+for (let x in author) {
+  text+= author[x] + "\n ";
+};
```
Comment (cjyuan) [inline Sprint-2/debug/author.js:21]: Can also explore using `Array.prototype.join()` to produce exactly the same string in one statement.

File: Sprint-2/implement/contains.test.js [hunk 6]
```diff
@@ -20,16 +20,57 @@ as the object doesn't contains a key of 'c'
...
+  expect(contains(obj, "d")).toBe(false);
+});
 
 // Given invalid parameters like an array
 // When passed to contains
 // Then it should return false or throw an error
+test("contains with invalid parameters return false", () => {
+  const obj = ["a", "b", "c", "d"];
+  expect(contains(obj, "a")).toBe(false);
+});
```
Comment (cjyuan) [inline Sprint-2/implement/contains.test.js:51]: Arrays are objects, with their indices acting as keys. That means

```javascript

  console.log("a" in ['a','b','c','d']);   // output false
  console.log("1" in ['a','b','c','d']);   // output true
```

A proper test should use a valid
key to ensure the function returns false specifically because the input is an array, not because the key is missing.

**Review by cjyuan — COMMENTED** [review]
Can you also address the two unaddressed comments from the previous review?

File: Sprint-2/implement/tally.js (hunk 4 above)
Comment (Mona-Eltantawy) [reply Sprint-2/implement/tally.js]: I changed it to `Object.hasOwn( )`.

**Comment by Mona-Eltantawy** [comment]
> Some of the code is not consistently formatted.
> 
> Have you installed the **prettier** VSCode extension and enabled "Format on save/paste" on VSCode, as recommended in https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/blob/main/readme.md ?
> 
> If you have enabled "Format on save" but it is not working, it is likely that you haven't assign a formatter for JS file. This could happen if you have zero or multiple extensions that can format .js file.
> 
> If you have installed "Prettier" extension. To assign it as the formatter of JS code, you can try:
> 
> 1. Use "Format document" to format the JS file. Sometimes, VSCode will ask you to choose a formatter, and you can manually select "Prettier".
> 2. Edit `settings.json` and set Prettier as the default formatter for JS.
>    See: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

I already install and use it.

File: Sprint-2/implement/contains.test.js (hunk 6 above)
Comment (Mona-Eltantawy) [reply Sprint-2/implement/contains.test.js:51]: I changed the test for the invalid paramater.

File: Sprint-2/debug/author.js [hunk 7]
```diff
@@ -11,6 +11,17 @@ const author = {
...
+console.log(text);
+
+// or: 
+const values = Object.values(author) .join ("");
```
Comment (cjyuan) [inline Sprint-2/debug/author.js]: You need to join using `"\n"` if you want each value to appear on a new line.

**Review by cjyuan — COMMENTED** [review]
Changes look good enough.

Can you address this commenet?
https://github.com/CodeYourFuture/Module-Data-Groups/pull/1195#discussion_r3077884558

File: Sprint-2/debug/author.js (hunk 7 above)
Comment (Mona-Eltantawy) [reply Sprint-2/debug/author.js]: do I have to do that If the function implementation already giving the expected output?

Comment (cjyuan) [reply Sprint-2/debug/author.js]: No need to change. I just need you to take note.

**Comment by cjyuan** [comment]
Can you address this commenet?
https://github.com/CodeYourFuture/Module-Data-Groups/pull/1195#discussion_r3077884558

File: Sprint-2/implement/contains.js (hunk 2 above)
Comment (Mona-Eltantawy) [reply Sprint-2/implement/contains.js:3]: as you said I can use either I kept the key in object in the function implemented already

**Comment by Mona-Eltantawy** [comment]
> Can you address this commenet? [#1195 (comment)](https://github.com/CodeYourFuture/Module-Data-Groups/pull/1195#discussion_r3077884558)

yeah I did, thank you for re-mention. .

**Comment by cjyuan** [comment]
All good now.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
